boa_ast/src/operations.rs

```diff
@@ -0,0 +1,261 @@
+//! Definitions of various **Syntax-Directed Operations** used in the [spec].
+//!
+//! [spec]: https://tc39.es/ecma262/#sec-syntax-directed-operations
+
+use core::ops::ControlFlow;
+
+use boa_interner::Sym;
+
+use crate::{
+    expression::{access::SuperPropertyAccess, Await, Identifier, SuperCall, Yield},
+    function::{
+        ArrowFunction, AsyncFunction, AsyncGenerator, Class, ClassElement, Function, Generator,
+    },
+    property::{MethodDefinition, PropertyDefinition},
+    visitor::{VisitWith, Visitor},
+    Expression,
+};
+
+/// Represents all the possible symbols searched for by the [`Contains`][contains] operation.
+///
+/// [contains]: https://tc39.es/ecma262/#sec-syntax-directed-operations-contains
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ContainsSymbol {
+    /// A node with the `super` keyword (`super(args)` or `super.prop`).
+    Super,
+    /// A super property access (`super.prop`).
+    SuperProperty,
+    /// A super constructor call (`super(args)`).
+    SuperCall,
+    /// A yield expression (`yield 5`).
+    YieldExpression,
+    /// An await expression (`await 4`).
+    AwaitExpression,
+    /// The new target expression (`new.target`).
+    NewTarget,
+    /// The body of a class definition.
+    ClassBody,
+    /// The super class of a class definition.
+    ClassHeritage,
+    /// A this expression (`this`).
+    This,
+    /// A method definition.
+    MethodDefinition,
+}
+
+/// Returns `true` if the node contains the given symbol.
+///
+/// More information:
+///  - [ECMAScript specification][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-static-semantics-contains
+#[must_use]
+pub fn contains<N>(node: &N, symbol: ContainsSymbol) -> bool
+where
+    N: VisitWith,
+{
+    struct ContainsVisitor(ContainsSymbol);
```

Comment:
I would at least derive `Debug`, `Clone`, and `Copy`. Also, we should add some documentation here. Maybe moving it out of the function could make sense, but this way is good too.

Reply:
Added the attributes and some docs on what the visitor does.
I'm more inclined on leaving it private to the function. Having both a visitor and a `contains` function in the same module that do the exact same thing could be a bit confusing.